tests/jobserver/test_pipeline.py

```diff
@@ -0,0 +1,27 @@
+import pytest
+
+from jobserver.pipeline import set_notifications_email
+
+from ..factories import UserFactory
+
+
+@pytest.mark.django_db
+def test_set_notifications_email_already_set():
+    user = UserFactory(
+        email="test@example.com", notifications_email="notifications@example.com"
+    )
+
+    set_notifications_email(user=user)
+
+    user.refresh_from_db()
+    assert user.notifications_email == "notifications@example.com"
+
+
+@pytest.mark.django_db
+def test_set_notifications_email_new_user():
+    user = UserFactory(email="test@example.com", notifications_email="")
+
+    set_notifications_email(user=user)
+
+    user.refresh_from_db()
+    assert user.notifications_email == "test@example.com"
```

Comment:
These tests give assurance that `set_notifications_email` works, but not that `set_notifications_email` ever gets called.  As such, they'd continue to pass even we upgrade Python Social Auth in such a way that means that `SOCIAL_AUTH_PIPELINE` is renamed (for instance).  Or if we migrate away from Python Social Auth (maybe we're only using 2% of the features but can't easily implement a new feature we need) then we'll have no tests to ensure that `notifications_email` is set.

It might not be possible to test this easily, in which case I wouldn't spend any time on it -- it's not mission critical and a blank `notifications_email` will (a) fail loudly and (b) be easily fixed.

Reply:
Very good point!  My only thought currently would be we'd have to mock/fake out the GitHub parts of the Login flow, starting with something like `client.get("/login")` and working from there to the same assert as above.

It's a lot of work, but you're right, we don't exercise the pipeline or login flow in that way at all, I think it might even handle #301 in a meaningful way too, and would allow checking things like #331.